CodingSession/ImageExtractinRTfield.py

```diff
@@ -152,30 +121,86 @@
+    where cast(Current_Case_Status__c as nvarchar(max)) is not null and cast(Current_Case_Status__c as nvarchar(max)) like '%</img>%'
+    union
+    select Id, CaseNumber, 'Conditions_of_Satisfaction__c', cast(Conditions_of_Satisfaction__c as nvarchar(max)), LastModifiedDate from SFCaseLand
+    where cast(Conditions_of_Satisfaction__c as nvarchar(max)) is not null and cast(Conditions_of_Satisfaction__c as nvarchar(max)) like '%</img>%') t
+    where t.LastModifiedDate >= ?
+""", lastrun)
+
+
+for row in cursor:
+    RTField_derived = {}
+    EID = row[0]
+    CaseNumber = row[1]
+    fieldname = row[2]
+    raw_text = row[3]
+
+    # find html tags with Salesforce link
+    new_text, MD5Hash_list = getsalesforceurlimage(raw_text)
+    #  find html tags with base64 encoded
+    getbase64encodedimage(new_text, MD5Hash_list)
+    #  find html tags with local file path embedded
+    image_url = re.findall('src="file:///C:\S+', row[3])  # remove the tag from rax text
+    if image_url:
+        pass
     #  Save the new text with its id and case number into a list of dictionary
     RTField_derived['CaseNumber'] = CaseNumber
     RTField_derived['Id'] = EID
     RTField_derived['APIName'] = fieldname
     RTField_derived['Rich_Text_Field_derived'] = new_text
     RTField_list.append(RTField_derived)
 
-# Is there a good way to do upsert? I cannot figure it out so I directly use truncate and load instead...
-# insert id, case number and new text into SQL table
-cursor.execute("truncate table SFRTFieldStage")
+
+#  insert id, case number and new text into SQL table
+#  cursor.execute("truncate table SFRTFieldStage")
 for each_RTField in RTField_list:
-    #  print(each_RTField)
-    # cursor.execute("update SFRTFieldStage set CaseNumber=?, Id=?,APIName=?,Rich_Text_Field_derived=?",
-    #               (each_RTField['CaseNumber'], each_RTField['Id'], each_RTField['APIName'], each_RTField['Rich_Text_Field_derived']))
-    cursor.execute("insert into SFRTFieldStage (CaseNumber, Id, APIName, Rich_Text_Field_derived) values (?, ?, ?, ?)",
-                   (each_RTField['CaseNumber'], each_RTField['Id'], each_RTField['APIName'],
-                    each_RTField['Rich_Text_Field_derived']))
+    cursor.execute("select * from SFRTFieldStage where Id=? and APIName=?",(each_RTField['Id'], each_RTField['APIName']))
+    if cursor.rowcount == 0:
+        cursor.execute("insert into SFRTFieldStage (CaseNumber, Id, APIName, Rich_Text_Field_derived) values (?, ?, ?, ?)",
+                       (each_RTField['CaseNumber'], each_RTField['Id'], each_RTField['APIName'], each_RTField['Rich_Text_Field_derived']))
+    else:
+        cursor.execute("update SFRTFieldStage set Rich_Text_Field_derived=? where Id=? and APIName=?",
+                       (each_RTField['Rich_Text_Field_derived'], each_RTField['Id'], each_RTField['APIName']))
     cursor.commit()
```

Comment:
NO! Never commit after EACH insert/update. This is SUPER slow.  Always commit in batches. Create a batch_size = 1000 (or whatever you like here) as a constant global variable at the top.
```
record_count = 0

loop:
   if record_count > 0 and record_count % batch_size = 0:
      cursor.commit()

   # insert/update

   record_count += 1

# After the loop, always commit at the end of the loop in case you're in the middle of a batch.
if record_count > 0
   cursor.commit()



```